design/rest-api-guideline.md

```diff
@@ -0,0 +1,476 @@
+# REST API Guideline
+
+* **Status**: Draft
+* **Author**: keycloak-team
+* **JIRA**: [KEYCLOAK-9344](https://issues.jboss.org/browse/KEYCLOAK-9344)
+
+## Abstract
+
+The Keycloak REST API Guideline provides a set of design principles and practices that should be considered by developers when designing, implementing and exposing a RESTful API. To provide the best experience for developers consuming Keycloak REST APIs, it's important API developers to follow to the principles and practices defined herein, making these APIs intuitive and easy to use.
+
+The goals of the guideline are:
+
+* Define the [overall principles](#principles) that should be taken into account when designing REST APIs for Keycloak
+
+* Define common practices and patterns for all Keycloak REST APIs such as:
+    * [Filtering](#filtering)
+    * [Pagination](#pagination)
+    * [Error Handling](#error-handling)
+    * [Documentation](#documentation)
+
+* Define common practices for [documentation](#documentation)
+
+Although the guideline is based on the REST architectural style, it does not try to strictly adhere to RESTful constraints. The principles covered by the guideline are based on practical experiences in the development of REST APIs in Keycloak while still conforming with most of the constraints of REST architectures.
+
+**Table of Contents**
+
+
+  - [Principles](#principles)
+    - [Level of Service Maturity](#level-of-service-maturity)
+    - [Resource Path](#resource-path)
+      - [Group](#group)
+      - [Versioning](#versioning)
+      - [Extensions](#extensions)
+      - [Technology Preview](#technology-preview)
+      - [Store Resources](#store-resources)
+      - [Controller Resources](#controller-resources)
+    - [Resource Representation](#resource-representation)
+    - [HTTP Status Codes](#http-status-codes)
+    - [HTTP Verbs](#http-verbs)
+      - [GET](#get)
+      - [POST](#post)
+      - [PUT](#put)
+      - [DELETE](#delete)
+  - [Practices](#practices)
+    - [Error Handling](#error-handling)
+    - [Filtering](#filtering)
+    - [Pagination](#pagination)
+    - [Concurrency Control and Consistency](#concurrency-control-and-concurrency)
```

Comment:
I would prefer to not remove it because:

* Document is marked as `Status: Draft`. I don't see an issue if we still have TODOs in a draft.
* Having them in the guideline helps to understand much better the document structure and what we want to cover in the document

Reply:
No problem, let's leave it as is.